introduce a concept of a reactor 'leader'

mattp- opened the pull request:
### What does this PR do?
in multimaster hot-hot environments, all events are streamed to all masters for a given minion. this introduces the problem of a reactor being executed on each hot master, regardless of if the reaction is idempotent or not. This patch introduces a simple is_leader bool that can be controlled via the reactor runner (or emitting the equivalent event to toggle) whether a reactor in hot-standby mode should be reacting to events coming into the bus. This provides a simple integration point for an external leader election tool such as consul-template. The feature itself is disconnected from multimaster, and is by default True, so it should be inert in most saltstack installs.

### What issues does this PR fix or reference?
n/a.

### Tests written?
Yes

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

mattp- committed:
introduce a concept of a reactor 'leader'

in multimaster hot-hot environments, all events are streamed to all masters for
a given minion. this introduces the problem of a reactor being executed on each
hot master, regardless of if the reaction is idempotent  or not. This patch
introduces a simple is_leader bool that can be controlled via the reactor
runner (or emitting the equivalent event to toggle) whether a reactor in
hot-standby mode should be reacting to events coming into the bus. The feature
itself is disconnected from multimaster, and is by default True, so it should
be inert in most saltstack installs. Tests forthcoming in following commit

mattp- commented:
hmm. it occurs to me with this PR that any minion can use this means of events that reactor/manage provides to interact (possibly maliciously) with a reactor. is this something that should be enabled by default? is there a safer means of interacting with master processes?

cachedout commented:
@mattp- That's my immediate concern as well. We should discuss this at length but one thing that comes to mind is potentially using the master key as an auth token, similar to the way that the LocalClient checks to ensure a request for a publication is coming from the master itself.

austinpapp commented:
but this is the general, any minion could list the reactors and delete all reactors. that would be the same behavior as emitting set leader.

mattp- committed:
Merge branch 'develop' into leader

mattp- committed:
Merge branch 'develop' into leader

cachedout commented:
@DmitryKuzmenko Could you please comment on this proposal?

DmitryKuzmenko commented on salt/utils/reactor.py:
```diff
@@ -263,13 +264,21 @@ def run(self):
+            if data['tag'].endswith('salt/reactors/manage/is_leader'):
+                self.event.fire_event({'result': self.is_leader},
+                                       'salt/reactors/manage/leader/value')
+            if data['tag'].endswith('salt/reactors/manage/set_leader'):
```
Let me to be a bit boring. =)
`elif`

DmitryKuzmenko commented on salt/utils/reactor.py:
```diff
@@ -263,13 +264,21 @@ def run(self):
+                    self.is_leader = data['data']['value']
+                self.event.fire_event({'result': self.is_leader},
+                                       'salt/reactors/manage/leader/value')
             if data['tag'].endswith('salt/reactors/manage/add'):
```
`elif`

DmitryKuzmenko reviewed (approved):
Approving with hope to see the requested changes applied.

DmitryKuzmenko commented:
Overall I'm OK with the proposed feature. It doesn't change the default behavior and if customers able to effectively use that leading driven reactions it's great.

mattp- commented:
great :+1: I still need to sort out one last piece which is adding the master key as an auth token, will do that before merge alongside your recommended changes. thanks!

cachedout commented:
Hi @mattp-  

Just checking with you. :) We're waiting on your additional changes before we do more reviews here and I wanted to make sure that matched your expectation. :)

mattp- committed:
Reactor: add master key auth to reactor runner

the event based api that was earlier implemented to allow for
controlling the reactor engine had no means of auth, any minion could
send in events to influence it. adding the master key ensures those
events come from the valid local master only.

mattp- committed:
Merge branch 'develop' into leader

mattp- commented:
@cachedout this should be good to merge now, I added the master key auth thing. That reactor ping test is spurious.
